File: frameworks/gin-rest/error_edge_cases_test.go
package main

import (
	"testing"

	"yourmodule/internal/models"
)

// ============================================================================
// Error: HTTP Status Codes
// ============================================================================

func TestHTTPStatusCodes(t *testing.T) {
	tests := []struct {
		name       string
		endpoint   string
		statusCode int
	}{
		{
			name:       "GET /users returns 200",
			endpoint:   "/api/users",
			statusCode: 200,
		},
		{
			name:       "GET /posts returns 200",
			endpoint:   "/api/posts",
			statusCode: 200,
		},
		{
			name:       "GET /nonexistent returns 404",
			endpoint:   "/api/nonexistent",
			statusCode: 404,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// In real test, would make HTTP request
			// Simulating success for list endpoints
			if tt.statusCode == 200 {
				// Success path
				t.Logf("Endpoint %s would return %d", tt.endpoint, tt.statusCode)
			} else {
				// Error path
				t.Logf("Endpoint %s would return %d", tt.endpoint, tt.statusCode)
			}
		})
	}
}

// ============================================================================
// Error: 404 Not Found
// ============================================================================

func TestNotFoundErrors(t *testing.T) {
	tests := []struct {
		name    string
		testId  string
	}{
		{
			name:   "user not found",
			testId: "nonexistent-user-id",
		},
		{
			name:   "post not found",
			testId: "nonexistent-post-id",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act
			if tt.name == "user not found" {
				user := factory.GetUser(tt.testId)
				if user != nil {
					t.Errorf("Expected nil for nonexistent user")
				}
			} else if tt.name == "post not found" {
				post := factory.GetPost(tt.testId)
				if post != nil {
					t.Errorf("Expected nil for nonexistent post")
				}
			}
		})
	}
}

// ============================================================================
// Error: Invalid Input
// ============================================================================

func TestInvalidInputHandling(t *testing.T) {
	tests := []struct {
		name       string
		inputType  string
	}{
		{
			name:      "invalid limit (negative)",
			inputType: "negative_limit",
		},
		{
			name:      "invalid limit (zero)",
			inputType: "zero_limit",
		},
		{
			name:      "very large limit",
			inputType: "large_limit",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			var limit int
			switch tt.inputType {
			case "negative_limit":
				limit = -5
			case "zero_limit":
				limit = 0
			case "large_limit":
				limit = 999999
			}

			// Act - Clamp limit
			result := limit
			if result < 0 {
				result = 0
			}
			if result > 100 {
				result = 100
			}

			// Assert
			if limit < 0 && result != 0 {
				t.Errorf("Negative limit should be clamped to 0")
			}
			if limit > 100 && result != 100 {
				t.Errorf("Large limit should be capped at 100")
			}
		})
	}
}

// ============================================================================
// Edge Case: UUID Validation
// ============================================================================

func TestUUIDValidation(t *testing.T) {
	tests := []struct {
		name     string
		testType string
	}{
		{
			name:     "all user IDs are UUID",
			testType: "user_uuid",
		},
		{
			name:     "all post IDs are UUID",
			testType: "post_uuid",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			validator := NewValidationHelper(t)
			defer factory.Reset()

			// Act & Assert
			if tt.testType == "user_uuid" {
				users := make([]*models.User, 3)
				for i := 0; i < 3; i++ {
					users[i] = factory.CreateTestUser(
						"user"+string(rune(48+i)),
						"user"+string(rune(48+i))+"@example.com",
						"User",
						"",
					)
					validator.AssertUUID(users[i].ID)
				}
			} else if tt.testType == "post_uuid" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				posts := make([]*models.Post, 3)
				for i := 0; i < 3; i++ {
					posts[i] = factory.CreateTestPost(author.ID, "Post "+string(rune(48+i)), "Content")
					validator.AssertUUID(posts[i].ID)
				}
			}
		})
	}
}

// ============================================================================
// Edge Case: Special Characters
// ============================================================================

func TestSpecialCharacterHandling(t *testing.T) {
	tests := []struct {
		name    string
		content string
	}{
		{
			name:    "single quotes",
			content: "I'm a developer",
		},
		{
			name:    "double quotes",
			content: `He said "hello"`,
		},
		{
			name:    "HTML tags",
			content: "Check <this> out",
		},
		{
			name:    "ampersand",
			content: "Tom & Jerry",
		},
		{
			name:    "emoji",
			content: "🎉 Celebration! 🚀 Rocket",
		},
		{
			name:    "accents",
			content: "Àlice Müller",
		},
		{
			name:    "diacritics",
			content: "José García",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act
			post := factory.CreateTestPost("user-id", "Title", tt.content)

			// Assert
			if post == nil {
				t.Errorf("Post creation failed")
				return
			}
			retrieved := factory.GetPost(post.ID)
			if retrieved == nil {
				t.Errorf("Post not found")
				return
			}
			if retrieved.Content != nil && *retrieved.Content != tt.content {
				t.Errorf("Content mismatch")
			}
		})
	}
}

// ============================================================================
// Edge Case: Boundary Conditions
// ============================================================================

func TestBoundaryConditions(t *testing.T) {
	tests := []struct {
		name    string
		length  int
		isValid bool
	}{
		{
			name:    "very long bio (5000 chars)",
			length:  5000,
			isValid: true,
		},
		{
			name:    "very long username (255 chars)",
			length:  255,
			isValid: true,
		},
		{
			name:    "very long post title",
			length:  500,
			isValid: true,
		},
		{
			name:    "very long content (5000 chars)",
			length:  5000,
			isValid: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			generator := NewDataGenerator(t)

			// Act
			content := generator.GenerateLongString(tt.length)

			// Assert
			if tt.isValid {
				if len(content) != tt.length {
					t.Errorf("Expected %d chars, got %d", tt.length, len(content))
				}
			}
		})
	}
}

// ============================================================================
// Edge Case: Null/Empty Fields
// ============================================================================

func TestNullEmptyFields(t *testing.T) {
	tests := []struct {
		name       string
		bio        string
		content    string
		wantNilBio bool
	}{
		{
			name:       "null bio is handled",
			bio:        "",
			content:    "",
			wantNilBio: true,
		},
		{
			name:       "empty string bio",
			bio:        "",
			content:    "Content",
			wantNilBio: true,
		},
		{
			name:       "present bio",
			bio:        "My bio",
			content:    "Content",
			wantNilBio: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			user := factory.CreateTestUser("alice", "alice@example.com", "Alice", tt.bio)

			// Assert
			if tt.wantNilBio {
				if user.Bio != nil {
					t.Errorf("Expected nil bio")
				}
			} else {
				if user.Bio == nil {
					t.Errorf("Expected non-nil bio")
				}
			}
		})
	}
}

// ============================================================================
// Edge Case: Relationship Validation
// ============================================================================

func TestRelationshipValidation(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "post author ID is valid UUID",
		},
		{
			name: "post references correct author",
		},
		{
			name: "multiple posts reference different authors",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			validator := NewValidationHelper(t)
			defer factory.Reset()

			if tt.name == "post author ID is valid UUID" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Post", "Content")

				// Assert
				validator.AssertUUID(post.AuthorID)
			} else if tt.name == "post references correct author" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Post", "Content")

				// Assert
				if post.AuthorID != author.ID {
					t.Errorf("Post does not reference correct author")
				}
			} else if tt.name == "multiple posts reference different authors" {
				author1 := factory.CreateTestUser("author1", "author1@example.com", "Author1", "")
				author2 := factory.CreateTestUser("author2", "author2@example.com", "Author2", "")

				post1 := factory.CreateTestPost(author1.ID, "Post1", "Content")
				post2 := factory.CreateTestPost(author2.ID, "Post2", "Content")

				// Assert
				if post1.AuthorID == post2.AuthorID {
					t.Errorf("Posts should reference different authors")
				}
				if post1.AuthorID != author1.ID || post2.AuthorID != author2.ID {
					t.Errorf("Posts do not reference correct authors")
				}
			}
		})
	}
}

// ============================================================================
// Edge Case: Data Type Validation
// ============================================================================

func TestDataTypeValidation(t *testing.T) {
	tests := []struct {
		name      string
		testType  string
	}{
		{
			name:     "username is string",
			testType: "string_username",
		},
		{
			name:     "post title is string",
			testType: "string_title",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act & Assert
			if tt.testType == "string_username" {
				user := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")
				if user.Username != "alice" {
					t.Errorf("Username not a string")
				}
			} else if tt.testType == "string_title" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Test Post", "Content")
				if post.Title != "Test Post" {
					t.Errorf("Title not a string")
				}
			}
		})
	}
}

// ============================================================================
// Edge Case: Uniqueness
// ============================================================================

func TestUniquenessConstraints(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "multiple users have unique IDs",
		},
		{
			name: "multiple posts have unique IDs",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			if tt.name == "multiple users have unique IDs" {
				user1 := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")
				user2 := factory.CreateTestUser("bob", "bob@example.com", "Bob", "")
				user3 := factory.CreateTestUser("charlie", "charlie@example.com", "Charlie", "")

				// Assert
				if user1.ID == user2.ID || user2.ID == user3.ID || user1.ID == user3.ID {
					t.Errorf("User IDs are not unique")
				}
			} else if tt.name == "multiple posts have unique IDs" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post1 := factory.CreateTestPost(author.ID, "Post1", "Content1")
				post2 := factory.CreateTestPost(author.ID, "Post2", "Content2")
				post3 := factory.CreateTestPost(author.ID, "Post3", "Content3")

				// Assert
				if post1.ID == post2.ID || post2.ID == post3.ID || post1.ID == post3.ID {
					t.Errorf("Post IDs are not unique")
				}
			}
		})
	}
}

File: frameworks/express-orm/tests/error-edge-cases.test.ts
import { TestFactory, ValidationHelper, DataGenerator } from './test-factory';

describe('Error and Edge Cases Tests', () => {
  let factory: TestFactory;

  beforeEach(() => {
    factory = new TestFactory();
  });

  // ============================================================================
  // Error: HTTP Status Codes
  // ============================================================================

  test('GET /users returns 200', () => {
    factory.createTestUser('alice', 'alice@example.com', 'Alice', '');
    expect(factory.getUserCount()).toBe(1);
  });

  test('GET /nonexistent returns 404', () => {
    const user = factory.getUser('nonexistent-id');
    expect(user).toBeUndefined();
  });

  // ============================================================================
  // Error: 404 Not Found
  // ============================================================================

  test('user not found returns undefined', () => {
    const user = factory.getUser('nonexistent-user-id');
    expect(user).toBeUndefined();
  });

  test('post not found returns undefined', () => {
    const post = factory.getPost('nonexistent-post-id');
    expect(post).toBeUndefined();
  });

  // ============================================================================
  // Error: Invalid Input
  // ============================================================================

  test('invalid limit (negative)', () => {
    const limit = -5;
    const clamped = Math.max(0, Math.min(100, limit));
    expect(clamped).toBe(0);
  });

  test('invalid limit (zero)', () => {
    const limit = 0;
    const clamped = Math.max(0, Math.min(100, limit));
    expect(clamped).toBe(0);
  });

  test('very large limit', () => {
    const limit = 999999;
    const clamped = Math.max(0, Math.min(100, limit));
    expect(clamped).toBe(100);
  });

  // ============================================================================
  // Edge Case: UUID Validation
  // ============================================================================

  test('all user IDs are UUID', () => {
    factory.createTestUser('user0', 'user0@example.com', 'User', '');
    factory.createTestUser('user1', 'user1@example.com', 'User', '');
    factory.createTestUser('user2', 'user2@example.com', 'User', '');

    factory.getAllUsers().forEach(user => {
      expect(() => ValidationHelper.assertUUID(user.id)).not.toThrow();
    });
  });

  test('all post IDs are UUID', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');

    factory.createTestPost(author.id, 'Post0', 'Content');
    factory.createTestPost(author.id, 'Post1', 'Content');
    factory.createTestPost(author.id, 'Post2', 'Content');

    factory.getAllPosts().forEach(post => {
      expect(() => ValidationHelper.assertUUID(post.id)).not.toThrow();
    });
  });

  // ============================================================================
  // Edge Case: Special Characters
  // ============================================================================

  test('single quotes', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', "I'm a developer");
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('double quotes', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', 'He said "hello"');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('HTML tags', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', 'Check <this> out');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('ampersand', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', 'Tom & Jerry');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('emoji', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', '🎉 Celebration! 🚀 Rocket');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('accents', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Àlice Müller', '');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  test('diacritics', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'José García', '');
    expect(factory.getUser(user.id)).toBeDefined();
  });

  // ============================================================================
  // Edge Case: Boundary Conditions
  // ============================================================================

  test('very long bio (5000 chars)', () => {
    const longBio = DataGenerator.generateLongString(5000);
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', longBio);
    const retrieved = factory.getUser(user.id);
    expect(retrieved?.bio?.length).toBe(5000);
  });

  test('very long username (255 chars)', () => {
    const longName = DataGenerator.generateLongString(255);
    const user = factory.createTestUser(longName, 'user@example.com', 'User', '');
    const retrieved = factory.getUser(user.id);
    expect(retrieved?.username.length).toBe(255);
  });

  test('very long post title', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const longTitle = DataGenerator.generateLongString(500);
    const post = factory.createTestPost(author.id, longTitle, 'Content');
    const retrieved = factory.getPost(post.id);
    expect(retrieved?.title.length).toBe(500);
  });

  test('very long content (5000 chars)', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const longContent = DataGenerator.generateLongString(5000);
    const post = factory.createTestPost(author.id, 'Title', longContent);
    const retrieved = factory.getPost(post.id);
    expect(retrieved?.content.length).toBe(5000);
  });

  // ============================================================================
  // Edge Case: Null/Empty Fields
  // ============================================================================

  test('null bio is handled', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', '');
    const retrieved = factory.getUser(user.id);
    expect(retrieved?.bio).toBeNull();
  });

  test('empty string bio', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', '');
    const retrieved = factory.getUser(user.id);
    expect(retrieved?.bio).toBeNull();
  });

  test('present bio', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', 'My bio');
    const retrieved = factory.getUser(user.id);
    expect(retrieved?.bio).toBe('My bio');
  });

  // ============================================================================
  // Edge Case: Relationship Validation
  // ============================================================================

  test('post author ID is valid UUID', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const post = factory.createTestPost(author.id, 'Post', 'Content');

    expect(() => ValidationHelper.assertUUID(author.id)).not.toThrow();
  });

  test('post references correct author', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const post = factory.createTestPost(author.id, 'Post', 'Content');

    expect(post.fk_author).toBe(author.pk_user);
  });

  test('multiple posts reference different authors', () => {
    const author1 = factory.createTestUser('author1', 'author1@example.com', 'Author1', '');
    const author2 = factory.createTestUser('author2', 'author2@example.com', 'Author2', '');

    const post1 = factory.createTestPost(author1.id, 'Post1', 'Content');
    const post2 = factory.createTestPost(author2.id, 'Post2', 'Content');

    expect(post1.fk_author).not.toBe(post2.fk_author);
    expect(post1.fk_author).toBe(author1.pk_user);
    expect(post2.fk_author).toBe(author2.pk_user);
  });

  // ============================================================================
  // Edge Case: Data Type Validation
  // ============================================================================

  test('username is string', () => {
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice', '');
    const retrieved = factory.getUser(user.id);
    expect(typeof retrieved?.username).toBe('string');
    expect(retrieved?.username).toBe('alice');
  });

  test('post title is string', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const post = factory.createTestPost(author.id, 'Test Post', 'Content');
    const retrieved = factory.getPost(post.id);
    expect(typeof retrieved?.title).toBe('string');
    expect(retrieved?.title).toBe('Test Post');
  });

  // ============================================================================
  // Edge Case: Uniqueness
  // ============================================================================

  test('multiple users have unique IDs', () => {
    const user1 = factory.createTestUser('alice', 'alice@example.com', 'Alice', '');
    const user2 = factory.createTestUser('bob', 'bob@example.com', 'Bob', '');
    const user3 = factory.createTestUser('charlie', 'charlie@example.com', 'Charlie', '');

    expect(user1.id).not.toBe(user2.id);
    expect(user2.id).not.toBe(user3.id);
    expect(user1.id).not.toBe(user3.id);
  });

  test('multiple posts have unique IDs', () => {
    const author = factory.createTestUser('author', 'author@example.com', 'Author', '');
    const post1 = factory.createTestPost(author.id, 'Post1', 'Content1');
    const post2 = factory.createTestPost(author.id, 'Post2', 'Content2');
    const post3 = factory.createTestPost(author.id, 'Post3', 'Content3');

    expect(post1.id).not.toBe(post2.id);
    expect(post2.id).not.toBe(post3.id);
    expect(post1.id).not.toBe(post3.id);
  });
});
